### DISCUSSION TITLE
A way to get the value that is used in a ternary operator without creating a new variable

### DISCUSSION THREAD

**Discussion by luizfernandonb** [dotnet/csharplang#4877] [Language Ideas]
Usually when we use a ternary operator, we want to return the value that was evaluated, and we usually create a variable that already stores this value, and using a variable to store a temporary value is unnecessary (at least in my view), so mine proposal is to create a keyword that takes the value that was analyzed in the ternary operator

Generally, we would do it like this:
```C#
var random = new Random();

int nextInt = random.Next();

Console.WriteLine(nextInt % 0 is 0 ? $"The number {nextInt} is an even number" : $"The number {nextInt} is an odd number");
```

What I want to propose is a keyword that prevents us from creating a new variable to store a value that we will compare and probably display later, as was done in the example above, thus saving time

According to what I thought, it would be something +- like:
```C#
var random = new Random();

Console.WriteLine(random.Next() % 0 is 0 ? $"The number {ternaryvalue} is an even number" : $"The number {ternaryvalue} is an odd number");
```

Don't mind the name `ternaryvalue`, it's just a symbol for what I mean, and also because I couldn't think of a better name, but anyway, it's pretty simple, if `random.Next()` generates any value, like 4318, instead of storing this value in another variable, the keyword `ternaryvalue` takes the value 4318 and displays it on the screen normally

I believe I have not been clear enough, any questions, just ask

**Comment by 333fred** [comment]
What you're looking for isn't just the value of the ternary, you're asking for the left-hand side of an expression inside the ternary, which is much more complex and not something we can really generally purpose in C#. You _could_ do something like this, but it's not elegant:

```cs
var random = new Random();
Console.WriteLine(random.Next is var nextInt && (nextInt % 0 is 0) ? $"The number {nextInt} is an even number" : $"The number {nextInt} is an odd number");
```

I honestly wouldn't recommend this though: your first example is much clearer.

**Reply by luizfernandonb** [reply]
> What you're looking for isn't just the value of the ternary, you're asking for the left-hand side of an expression inside the ternary, which is much more complex and not something we can really generally purpose in C#. You _could_ do something like this, but it's not elegant:
> 
> ```cs
> var random = new Random();
> Console.WriteLine(random.Next is var nextInt && (nextInt % 0 is 0) ? $"The number {nextInt} is an even number" : $"The number {nextInt} is an odd number");
> ```
> 
> I honestly wouldn't recommend this though: your first example is much clearer.

I didn't know it could be done like that you showed, and really, it's not very clear, my proposal could even simplify your code more, if it wasn't so complicated for C#

**Reply by jnm2** [reply]
> What you're looking for isn't just the value of the ternary, you're asking for the left-hand side of an expression inside the ternary, which is much more complex and not something we can really generally purpose in C#.

I agree, and it's even worse. In Luiz's example, `random.Next() % 0` is what `ternaryvalue` would refer to. C# has no way to know that it should stop at `random.Next()` instead. It's much better to be explicit.

(Also, `% 0` is division by zero. I'm guessing `% 2` was meant.)

**Reply by luizfernandonb** [reply]
> > What you're looking for isn't just the value of the ternary, you're asking for the left-hand side of an expression inside the ternary, which is much more complex and not something we can really generally purpose in C#.
> 
> I agree, and it's even worse. In Luiz's example, `random.Next() % 0` is what `ternaryvalue` would refer to. C# has no way to know that it should stop at `random.Next()` instead. It's much better to be explicit.
> 
> (Also, `% 0` is division by zero. I'm guessing `% 2` was meant.)

This was the only example I thought of, but it works for any other example that needs to get the same value being compared in the ternary operator

**Reply by CyrusNajmabadi** [reply]
> the same value being compared in the ternary operator

the value for a ternary will always be either `true` or `false`. you are grabbing some other value based on criteria you havent specified yet :-)

**Reply by luizfernandonb** [reply]
> > the same value being compared in the ternary operator
> 
> the value for a ternary will always be either `true` or `false`. you are grabbing some other value based on criteria you havent specified yet :-)

What I'm specifying is to get the value that's being compared to anything else in the ternary operator, in the example I gave, it's `random.Next()`, but it could be anything else

**Reply by jnm2** [reply]
The value that is being compared in the example you gave is not `random.Next()`, it's `random.Next() % 0`. You can tell because `%` is not a comparison operator.

**Reply by luizfernandonb** [reply]
> The value that is being compared in the example you gave is not `random.Next()`, it's `random.Next() % 0`. You can tell because `%` is not a comparison operator.

You're right, I think we've reached the point @333fred said, that it would be complicated to implement this in C#

**Reply by luizfernandonb** [reply]
What I ultimately want is just to get `random.Next()`, but apparently it's pretty impossible

**Reply by jnm2** [reply]
No possibility stands out to me that would not also be a super subtle set of rules, making it hard to understand when reading the code. That's why explicit variable creation feels better even though you have to type more and invent a name to describe the value you're capturing.

**Reply by CyrusNajmabadi** [reply]
random.Next() is var ternvalue && ...

**Reply by theunrepentantgeek** [reply]
@luizfernandonb wrote
> What I ultimately want is just to get random.Next(), but apparently it's pretty impossible

It's not impossible - and you don't need to wait years for a new, complicated, hard-to-predict, impossible-to-discover, and difficult-to-document feature to be debated, relitigated, championed, designed, implemented, tested, and released.

``` csharp
int nextInt = random.Next();
```

_Using a temporary variable is the right solution._

The LDM have been quite clear in the past that making C# "more concise" is an explicit non-goal. 
Saving keystrokes at the cost of impairing readability is a step backwards, not forwards.

**Comment by michaelcheers** [comment]
```csharp
var random = new Random();
Console.WriteLine((defer random.Next()) % 2 == 0 ? $"The number {it} is an even number" : $"The number {it} is an odd number");
```

**Reply by luizfernandonb** [reply]
`defer` would be what exactly?

**Reply by michaelcheers** [reply]
it would set the variable 'it' to the expression.

**Reply by luizfernandonb** [reply]
> it would set the variable 'it' to the expression.

Ok, it's because the name `it` reminds me of iterators that C++ has

**Reply by HaloFour** [reply]
I don't understand what `defer` is supposed to mean here.  What is being deferred?  What does that have to do with ternary operations and why does it introduce a new identifier into scope?  How would those identifiers be nested?

I'm not a fan of a specific C# syntax to enable what OP is suggesting.  It's simply much less clear what it's doing.  If you need to perform two operations on the value (which is what is happening) then a temporary local variable isn't just appropriate, it's idiomatic.

**Reply by CyrusNajmabadi** [reply]
Defer makes me think: happens at the end.

**Reply by jnm2** [reply]
`capture` instead of `defer`?

If it was `capture`...`@`, wanting `(capture random.Next()) % 2 == 0 ? $"The number {@} is an even number" : ...` reminds me vaguely of wanting `people.Where(@Age >= 65).Select(@Name)`.

**Reply by michaelcheers** [reply]
How about

```csharp
with(random.Next())(@ % 2 == 0 ? $"The number {@} is an even number" : $"The number {@} is an odd number")
```

Or similarly:

```csharp
(v => v % 2 == 0 ? $"The number {v} is an even number" : $"The number {v} is an odd number")(random.Next())
```

which is also not currently valid but is in javascript.

**Reply by huoyaoyuan** [reply]
> Or similarly:
> 
> ```cs
> (v => v % 2 == 0 ? $"The number {v} is an even number" : $"The number {v} is an odd number")(random.Next())
> ```
> 
> which is also not currently valid but is in javascript.

This should be valid soon with lambda improvement feature.

**Reply by jnm2** [reply]
I thought you would have to do `((int v) => ...)(random.Next())`.

**Comment by huoyaoyuan** [comment]
> using a variable to store a temporary value is unnecessary

I would say that using a variable is necessary. If you are defining much more variables, you should consider creating a helper method.

And for this specific case, `random.Next()` has side effect and I will avoid to call it in nested expression.
